src/transformers/modeling_blenderbot.py

```diff
@@ -0,0 +1,582 @@
+    """
+    implements a blenderbot generator model with an encoder and decoder
+    """
+    def __init__(self, config):
+        self.pad_idx = config.pad_idx
+        self.start_idx = config.start_idx
+        self.end_idx = config.end_idx
+        super().__init__(config)
+        self.embeddings = nn.Embedding(config.vocab_size, config.embedding_size, self.pad_idx)
+        self.n_positions = config.n_positions if config.n_positions > 0 else 1024
+        self.encoder = BlenderbotEncoder(
+            n_heads=config.n_heads,
+            n_layers=config.n_layers,
+            embedding_size=config.embedding_size,
+            ffn_size=config.ffn_size,
+            vocabulary_size=config.vocab_size,
+            embedding=self.embeddings,
+            dropout=config.dropout,
+            relu_dropout=config.relu_dropout,
+            attention_dropout=config.attention_dropout,
+            padding_idx=self.pad_idx,
+            learn_positional_embeddings=config.learn_positional_embeddings,
+            n_positions=self.n_positions,
+            activation=config.activation,
+        )
+        self.decoder = BlenderbotDecoder(
+            n_heads=config.n_heads,
+            embedding_size=config.embedding_size,
+            ffn_size=config.ffn_size,
+            n_layers=config.n_layers,
+            vocabulary_size=config.vocab_size,
+            embedding=self.embeddings,
+            attention_dropout=config.attention_dropout,
+            dropout=config.dropout,
+            relu_dropout=config.relu_dropout,
+            learn_positional_embeddings=config.learn_positional_embeddings,
+            padding_idx=self.pad_idx,
+            n_positions=self.n_positions,
+            activation=config.activation
+        )
+        self.init_weights()
+
+    def forward(self, input_ids, decoder_input_ids=None, position_ids=None, token_type_ids=None):
+        if input_ids is None:
+            raise ValueError("Your input tensor is None, you should specify it")
+        encoder_out = self.encoder(input_ids, token_type_ids)
+        assert isinstance(encoder_out, tuple)
+
+        def prepare_decoder_inputs(input_ids):
+            prev_output_tokens = input_ids.clone()
+            index_of_eos = (input_ids.ne(self.end_idx).sum(dim=1) - 1).unsqueeze(-1)
+            prev_output_tokens[:, 0] = input_ids.gather(1, index_of_eos).squeeze()
+            prev_output_tokens[:, 1:] = input_ids[:, :-1]
+            return prev_output_tokens
+        if decoder_input_ids is None:
+            decoder_input_ids = prepare_decoder_inputs(input_ids)
+        decoder_output = self.decoder(decoder_input_ids, encoder_out)
+        assert isinstance(decoder_output, tuple)
+        tensor_out, new_inc_state = decoder_output
+        output = F.linear(tensor_out, self.embeddings.weight)
+        # force the start token  probability of generation to be 0.
+        output[:, :, self.start_idx] = neginf(output.dtype)
+        return output
```

Comment:
This looks good to me, but it would be nice if @patrickvonplaten can have a look at it and check that it matches the recommended practices for encoder-decoder models in the library at the moment.